File: src/lib/sabflow/forge/blocks/n8n/hr/twist.ts
/**
 * Forge block: Twist
 *
 * Source: n8n-master/packages/nodes-base/nodes/Twist/Twist.node.ts
 *
 * Auth: Bearer accessToken on `https://api.twist.com/api/v3`.
 *
 * Operations covered:
 *   - thread.add      POST /threads/add
 *   - channel.list    GET  /channels/get
 *   - comment.add     POST /comments/add
 */

import { registerForgeBlock } from '../../../registry';
import type {
  ForgeActionContext,
  ForgeActionResult,
  ForgeBlock,
} from '../../../types';
import { apiRequest, asString } from '../_shared/http';

const API = 'https://api.twist.com/api/v3';

function authHeaders(ctx: ForgeActionContext): Record<string, string> {
  const token = asString(ctx.options.accessToken);
  if (!token) throw new Error('Twist: accessToken is required');
  return { Authorization: `Bearer ${token}`, Accept: 'application/json' };
}

async function threadAdd(ctx: ForgeActionContext): Promise<ForgeActionResult> {
  const channelId = asString(ctx.options.channelId);
  const title = asString(ctx.options.title);
  const content = asString(ctx.options.content);
  if (!channelId || !title || !content) {
    throw new Error('Twist: channelId, title and content are required');
  }
  const body = {
    channel_id: Number(channelId),
    title,
    content,
  };
  const res = await apiRequest({
    service: 'Twist',
    method: 'POST',
    url: `${API}/threads/add`,
    headers: authHeaders(ctx),
    json: body,
  });
  return { outputs: { thread: res.data }, logs: [`Twist thread add → ${title}`] };
}

async function channelList(ctx: ForgeActionContext): Promise<ForgeActionResult> {
  const workspaceId = asString(ctx.options.workspaceId);
  if (!workspaceId) throw new Error('Twist: workspaceId is required');
  const params = new URLSearchParams({ workspace_id: workspaceId });
  const res = await apiRequest({
    service: 'Twist',
    method: 'GET',
    url: `${API}/channels/get?${params.toString()}`,
    headers: authHeaders(ctx),
  });
  return { outputs: { channels: res.data }, logs: [`Twist channel list → ${workspaceId}`] };
}

async function commentAdd(ctx: ForgeActionContext): Promise<ForgeActionResult> {
  const threadId = asString(ctx.options.threadId);
  const content = asString(ctx.options.content);
  if (!threadId || !content) throw new Error('Twist: threadId and content are required');
  const body = {
    thread_id: Number(threadId),
    content,
  };
  const res = await apiRequest({
    service: 'Twist',
    method: 'POST',
    url: `${API}/comments/add`,
    headers: authHeaders(ctx),
    json: body,
  });
  return { outputs: { comment: res.data }, logs: [`Twist comment add → ${threadId}`] };
}

const block: ForgeBlock = {
  id: 'forge_twist',
  name: 'Twist',
  description: 'Post threads, list channels and add comments in Twist.',
  iconName: 'LuMessageSquare',
  category: 'Integration',
  auth: { type: 'none' },
  actions: [
    {
      id: 'thread_add',
      label: 'Add thread',
      description: 'Start a new thread in a channel.',
      fields: [
        { id: 'accessToken', label: 'Access token', type: 'password', required: true },
        { id: 'channelId', label: 'Channel ID', type: 'text', required: true },
        { id: 'title', label: 'Title', type: 'text', required: true },
        { id: 'content', label: 'Content', type: 'text', required: true },
      ],
      run: threadAdd,
    },
    {
      id: 'channel_list',
      label: 'List channels',
      description: 'List channels in a workspace.',
      fields: [
        { id: 'accessToken', label: 'Access token', type: 'password', required: true },
        { id: 'workspaceId', label: 'Workspace ID', type: 'text', required: true },
      ],
      run: channelList,
    },
    {
      id: 'comment_add',
      label: 'Add comment',
      description: 'Add a comment to a thread.',
      fields: [
        { id: 'accessToken', label: 'Access token', type: 'password', required: true },
        { id: 'threadId', label: 'Thread ID', type: 'text', required: true },
        { id: 'content', label: 'Content', type: 'text', required: true },
      ],
      run: commentAdd,
    },
  ],
};

registerForgeBlock(block);
export default block;

File: rust/crates/sabflow-nodes/src/nodes/twist.rs
//! Twist node — async team messaging (Doist).
//!
//! Posts messages and comments via the Twist API v3
//! (`https://api.twist.com/api/v3`). Auth is bearer token (OAuth2 access
//! token stored in the credential `accessToken` field).
//!
//! Quality bar: C.5 typed-stub-with-descriptor — `addMessage` and
//! `addComment` are wired against the live API; other resources follow.

use async_trait::async_trait;
use serde_json::{Value, json};

use crate::{
    context::{ExecutionContext, NodeInput, NodeOutput},
    descriptor::{
        CredentialBinding, NodeCategory, NodeDescriptor, NodeProperty, NodePropertyOption,
        NodePropertyType,
    },
    error::{NodeError, NodeResult},
    node::Node,
};

const TWIST_API: &str = "https://api.twist.com/api/v3";

pub struct TwistNode;

fn opt(name: &str, value: &str) -> NodePropertyOption {
    NodePropertyOption {
        name: name.to_string(),
        value: json!(value),
        description: None,
    }
}

#[async_trait]
impl Node for TwistNode {
    fn descriptor(&self) -> NodeDescriptor {
        NodeDescriptor::new(
            "twist",
            "Twist",
            "Post threads and comments to Twist channels",
            NodeCategory::Communication,
        )
        .icon("message-square")
        .color("#EC6E5E")
        .credentials(vec![CredentialBinding {
            name: "twistOAuth2Api".into(),
            display_name: "Twist OAuth2 Access Token".into(),
            required: true,
        }])
        .properties(vec![
            NodeProperty::new("credentialId", "Credential", NodePropertyType::Credential)
                .required(),
            NodeProperty::new("resource", "Resource", NodePropertyType::Options)
                .options(vec![
                    opt("Thread", "thread"),
                    opt("Comment", "comment"),
                ])
                .default(json!("thread"))
                .required(),
            NodeProperty::new("operation", "Operation", NodePropertyType::Options)
                .options(vec![opt("Add", "add")])
                .default(json!("add"))
                .required(),
            NodeProperty::new("channelId", "Channel ID", NodePropertyType::String)
                .description("Twist channel ID where the thread is posted")
                .show_when("resource", &["thread"])
                .required(),
            NodeProperty::new("title", "Title", NodePropertyType::String)
                .placeholder("Weekly update")
                .show_when("resource", &["thread"])
                .required(),
            NodeProperty::new("threadId", "Thread ID", NodePropertyType::String)
                .description("Twist thread ID to comment on")
                .show_when("resource", &["comment"])
                .required(),
            NodeProperty::new("content", "Content", NodePropertyType::String)
                .placeholder("Hello team!")
                .description("Message body — Markdown supported")
                .required(),
        ])
    }

    async fn execute(
        &self,
        ctx: &mut ExecutionContext,
        _input: NodeInput,
        params: &Value,
    ) -> NodeResult<NodeOutput> {
        let cred_id = ctx.param_str(params, "credentialId")?;
        let cred = ctx.credential(&cred_id)?;
        let token = cred
            .data
            .get("accessToken")
            .and_then(|v| v.as_str())
            .ok_or_else(|| NodeError::MissingParameter("accessToken".into()))?
            .to_string();

        let resource = ctx
            .param_str_opt(params, "resource")
            .unwrap_or_else(|| "thread".to_string());
        let operation = ctx
            .param_str_opt(params, "operation")
            .unwrap_or_else(|| "add".to_string());
        let content = ctx.param_str(params, "content")?;

        let (path, payload) = match (resource.as_str(), operation.as_str()) {
            ("thread", "add") => {
                let channel_id = ctx.param_str(params, "channelId")?;
                let title = ctx.param_str(params, "title")?;
                (
                    "/threads/add",
                    json!({
                        "channel_id": channel_id,
                        "title": title,
                        "content": content,
                    }),
                )
            }
            ("comment", "add") => {
                let thread_id = ctx.param_str(params, "threadId")?;
                (
                    "/comments/add",
                    json!({
                        "thread_id": thread_id,
                        "content": content,
                    }),
                )
            }
            (res, op) => {
                return Err(NodeError::InvalidParameter {
                    name: "operation".into(),
                    reason: format!("unsupported twist {res} operation: {op}"),
                });
            }
        };

        let url = format!("{TWIST_API}{path}");
        let res = ctx
            .http
            .post(&url)
            .bearer_auth(&token)
            .header("Content-Type", "application/json")
            .json(&payload)
            .send()
            .await?;
        let status = res.status();
        let text_body = res.text().await.unwrap_or_default();
        let body: Value =
            serde_json::from_str(&text_body).unwrap_or(Value::String(text_body.clone()));
        if !status.is_success() {
            return Err(NodeError::UpstreamError {
                status: status.as_u16(),
                body: body.to_string(),
            });
        }
        Ok(NodeOutput::single(vec![body]))
    }
}
